File: training/models/cfc_cell.py
"""Closed-form Continuous-time cell (Hasani et al.).

Input-dependent time constants learn when to update vs retain memories.
Drop-in replacement for TemporalGRU with the same forward signature.
"""

from __future__ import annotations

import torch
import torch.nn as nn

CFC_H_DIM = 256  # default CfC hidden dimension (wider than GRU for richer temporal state)


class CfCCell(nn.Module):
    """Closed-form Continuous-time cell (Hasani et al.).

    Input-dependent time constants learn when to update vs retain memories.
    Drop-in replacement for TemporalGRU with the same forward signature.
    """

    def __init__(self, d_model: int, h_dim: int = CFC_H_DIM):
        super().__init__()
        self.h_dim = h_dim
        total = d_model + h_dim
        self.f_gate = nn.Linear(total, h_dim)
        self.h_gate = nn.Linear(total, h_dim)
        self.t_a = nn.Linear(total, h_dim)
        self.t_b = nn.Linear(total, h_dim)
        self.proj = nn.Linear(h_dim, d_model)

        # Init: start near full memory retention
        nn.init.constant_(self.f_gate.bias, 1.0)
        nn.init.constant_(self.t_b.bias, 1.0)

    def forward(
        self, x: torch.Tensor, h_prev: torch.Tensor | None = None,
        delta_t: float = 1.0,
    ) -> tuple[torch.Tensor, torch.Tensor]:
        """
        Args:
            x: (B, d_model) -- pooled entity representation for this tick
            h_prev: (B, h_dim) -- hidden state from previous tick, or None
            delta_t: time scaling factor (1.0 = standard tick interval)
        Returns:
            projected: (B, d_model) -- temporally enriched representation
            h_new: (B, h_dim) -- hidden state to propagate
        """
        if h_prev is None:
            h_prev = torch.zeros(x.shape[0], self.h_dim, device=x.device)
        combined = torch.cat([x, h_prev], dim=-1)
        f = torch.sigmoid(self.f_gate(combined))
        candidate = torch.tanh(self.h_gate(combined))
        t = torch.sigmoid(self.t_a(combined)) * delta_t + self.t_b(combined)
        h_new = torch.tanh(f * h_prev + (1 - f) * candidate * t)
        return self.proj(h_new), h_new

File: crates/tactical_sim/src/sim/burn_model/cfc_cell.rs
//! Closed-form Continuous-time (CfC) temporal cell.
//!
//! Replaces GRU for temporal context. Input-dependent time constants
//! allow the cell to learn different integration speeds for different
//! game states.

use burn::module::Module;
use burn::nn::{Linear, LinearConfig};
use burn::prelude::*;
use burn::tensor::activation::sigmoid;

#[derive(Module, Debug)]
pub struct CfCCell<B: Backend> {
    f_gate: Linear<B>,  // (d_model + h_dim) -> h_dim
    h_gate: Linear<B>,  // (d_model + h_dim) -> h_dim
    t_a: Linear<B>,     // (d_model + h_dim) -> h_dim
    t_b: Linear<B>,     // (d_model + h_dim) -> h_dim
    proj: Linear<B>,    // h_dim -> d_model
    h_dim: usize,
}

#[derive(Config, Debug)]
pub struct CfCCellConfig {
    pub d_model: usize,
    #[config(default = "64")]
    pub h_dim: usize,
}

impl CfCCellConfig {
    pub fn init<B: Backend>(&self, device: &B::Device) -> CfCCell<B> {
        let input_dim = self.d_model + self.h_dim;
        CfCCell {
            f_gate: LinearConfig::new(input_dim, self.h_dim).init(device),
            h_gate: LinearConfig::new(input_dim, self.h_dim).init(device),
            t_a: LinearConfig::new(input_dim, self.h_dim).init(device),
            t_b: LinearConfig::new(input_dim, self.h_dim).init(device),
            proj: LinearConfig::new(self.h_dim, self.d_model).init(device),
            h_dim: self.h_dim,
        }
    }
}

impl<B: Backend> CfCCell<B> {
    pub fn h_dim(&self) -> usize {
        self.h_dim
    }

    /// CfC forward: (x, h_prev, delta_t) -> (output, h_new)
    ///
    /// x: [B, d_model], h_prev: [B, h_dim], delta_t: scalar
    pub fn forward(
        &self,
        x: Tensor<B, 2>,
        h_prev: Tensor<B, 2>,
        delta_t: f32,
    ) -> (Tensor<B, 2>, Tensor<B, 2>) {
        // combined = cat([x, h_prev])
        let combined = Tensor::cat(vec![x, h_prev.clone()], 1);

        // f = sigmoid(f_gate(combined))
        let f = sigmoid(self.f_gate.forward(combined.clone()));

        // candidate = tanh(h_gate(combined))
        let candidate = self.h_gate.forward(combined.clone()).tanh();

        // t = sigmoid(t_a(combined)) * delta_t + t_b(combined)
        let t_a_out = sigmoid(self.t_a.forward(combined.clone()));
        let t_b_out = self.t_b.forward(combined);
        let t = t_a_out * delta_t + t_b_out;

        // h_new = tanh(f * h_prev + (1 - f) * candidate * t)
        let ones = f.clone().ones_like();
        let h_new = (f.clone() * h_prev + (ones - f) * candidate * t).tanh();

        // output = proj(h_new)
        let output = self.proj.forward(h_new.clone());

        (output, h_new)
    }
}
